[inductor] comprehensive padding

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120758

This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang @jansel @Chillee @eellison

shunting314 committed:
[WIP][inductor] padding

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison reviewed (commented):
Cool!

- We might need some logic around not increasing memory too much. Not sure if this would actually be an issue in practice.
- I think we still want to preserve strides for user visible outputs.
- We should make sure that we're still appropriately passing in dense tensors to extern ops that require it.. I think this will just work but worth double checking. 
- Not sure if there is an aot autograd related issue of changing strides... I dont think there is but I remember it has special logic to deal with changing strides from inductor

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 commented:
Current result [link](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2012%20Mar%202024%2017%3A40%3A05%20GMT&stopTime=Tue%2C%2019%20Mar%202024%2017%3A40%3A05%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/103/head&lCommit=5614f05f8ebf8115e94d1e7acfb58649add0daf9&rBranch=main&rCommit=65ccac6f175912062926886ebc86ee6b09a7ce23)
- HF 4% speedup
- TIMM/TB 4% slowdown

I debugged the slowdown for nvidia_deeprecommender. The root cause is we change the stride of variable/gradients which can result in an extra copy in AccumulateGrad. The point of the copy is to make sure variable and its gradient have the same layout so optimizer can be more efficient. But with tiling, a compiled optimizer does not really benefit a lot from that. I added an option to skip the extra copy in AccumulateGrad and bring back the perf for nvidia_deeprecommender. Will try other slowed models. Perf for nvidia_deeprecommender:
- no padding: 10.9ms
- with padding: 11.68ms
- with padding and skip gradient layout contract: 10.63ms. It brings back the perf and shows some speedup.

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

Chillee commented:
I'm pretty skeptical about adding a mode to autograd that relaxes this restriction 🤔 Why can't you just disable the layout optimization for gradient computation?

shunting314 commented:
> I'm pretty skeptical about adding a mode to autograd that relaxes this restriction

What's the concern here? From the code comment, the grad layout contract is mainly to make the optimizer (and c10d reducer as mentioned by the comment. But I'm not sure if that's very relevant here) efficient. But with an compiled optimizer, let variable and gradient having different layout is not a big problem.

> Why can't you just disable the layout optimization for gradient computation?

I think that's possible. I'm just worried that if we restrict too much on what tensors we can change the layout, we would get less speedup. But if we think keeping gradients' layout the same as eager is important, we can do that. On the other hand, I need a way to figure out what tensors in the bwd graph are gradients.

Chillee commented:
> On the other hand, I need a way to figure out what tensors in the bwd graph are gradients.

All the outputs of the backwards graph are gradients.

albanD commented:
Chiming in here quickly. Disabling the grad layout is indeed a bit problematic.
In general, we aim to preserve layout of the objects going back to the user in compile. And for AccumulateGrad that has a very complex semantic, we went through quite a bit of trouble to use the original function instead of trying to duplicate to guarantee exact 1-1 correspondence between the two.
This flag would definitely defeat that and if we're going down this path of breaking the layout preservation, we should just have accumulate grad done directly within compile and not re-use accumulate grad at all instead of adding a flag there.

That being said, I don't think you actually want to break that contract?
One option depending on your use case is that if the .grad field is already allocated, we will re-use at as-is though. So you can set the .grad field ahead of time (or zero it out instead of setting it to None between allocations) to preserve a given layout there.

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 commented:
<img width="487" alt="Screenshot 2024-04-05 at 10 35 33 AM" src="https://github.com/pytorch/pytorch/assets/52589240/82209830-b299-4e83-b44d-1a279d001846">

Perf with cudagraphs looks good now. The perf without cudagraps for TB still has large regress though

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 commented:
<img width="497" alt="Screenshot 2024-04-06 at 10 21 08 PM" src="https://github.com/pytorch/pytorch/assets/52589240/b7cc5cff-2135-4c5c-bf9c-b07702b4bf7e">

All perf regress has been fixed now. It turns out that the large regress from default setting from TB is due to using a new triton pin. It seems that triton regresses.. Reverting the pin change eventually fixes it. But this also means we are not using Hongtao's fix here: https://github.com/openai/triton/pull/3497 (since I need use a relatively new triton to apply the PR). I'll try to dig more.

On the other hand, I'll start cleaning up this PR to make it ready to review.

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] padding"


This is not fully ready for review yet. Sending it out in case people want to give early feedback. Also it can allow me run CI perf test earlier.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup. I feel the solution can be further improved to get more speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

shunting314 commented:
@jansel , @Chillee , @eellison  just a heads up that this PR is ready for review.

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -322,6 +322,22 @@ def is_tf32_warning_applicable(gm: torch.fx.GraphModule):
+    )
+
+    if config.comprehensive_padding and not has_cuda:
+        perf_hint_log.warning("Skip comprehensive padding on CPU")
```
This doesn't seem like it should be a warning.

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -1344,6 +1364,15 @@ def partition_fn(graph, joint_inputs, **kwargs):
+            *_, model_outputs_node = model.graph.nodes
+            assert model_outputs_node.op == "output"
+            model_outputs = pytree.arg_tree_leaves(*model_outputs_node.args)
+            user_visible_outputs = {
+                n.name for n in model_outputs if isinstance(n, torch.fx.Node)
+            }
```
sets have non-determinstic order, which can lead to annoying issues. Suggest `dict.fromkeys`

jansel commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+            return in_strides
+
+        # a loose check for channels last
+        if not config.pad_channels_last and len(in_strides) == 4 and in_strides[1] == 1:
```
I think there is a helper elsewhere in this file for this check, reuse or refactor it?

jansel commented on torch/_inductor/metrics.py:
```diff
@@ -43,6 +43,8 @@
 # Each element counts the number of inner kernels in each outer loop fusion.
 cpp_outer_loop_fused_inner_counts: List[int] = []
 
+num_comprehensive_padding = 0
```
add to `reset()`,etc below

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -336,7 +352,9 @@ def count_bytes_inner(
-    with V.set_graph_handler(graph), V.set_real_inputs(example_inputs):
+    with V.set_graph_handler(graph), V.set_real_inputs(
+        example_inputs
+    ), maybe_disable_comprehensive_padding(example_inputs):
```
any reason to do this as a config patch instead of as a field on GraphLowering?

shunting314 replied:
Nope, either way should work. I can access the config by either config.comprehensive_padding or V.graph.comprehensive_padding .

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -1344,6 +1364,15 @@ def partition_fn(graph, joint_inputs, **kwargs):
+        user_visible_outputs = set()
+
+        if config.bw_outputs_user_visible:
+            *_, model_outputs_node = model.graph.nodes
```
nit: `torch._inductor.utils.output_node`

eellison commented on torch/_inductor/config.py:
```diff
@@ -421,6 +421,16 @@ def decide_compile_threads():
+
+# Whether to treat output of the backward graph as user visible.
+# For user visible outputs, inductor will make sure the stride matches with eager.
+bw_outputs_user_visible = True
```
Can you say more why you needed to do this ? I saw alban comment something along the lines of gradient stealing or something?

shunting314 replied:
Previously inductor does not guarantee that the output of backward graph (e.g. gradients for parameters or graph inputs if they require gradients) have the same strides as eager. This may ends up with the case that a gradient has a different layout compared to the corresponding parameter. The PyTorch autograd engine has something called gradient layout contract that may do a extra copy in that case so that the strides of gradient and parameter match. This makes the optimizer more efficient. Treating output of backward graph user visible can avoid that extra copy.

On the other hand, it's just good to make the strides of gradients matching eager in case user runs some view op on them.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -148,6 +149,61 @@ def getattr_recursive(obj, target):
+        aten.convolution_backward,
+    }
+    # what's a better way to collect the reduction ops?
+    ops_like_padding = {
```
It would be great (not needed for this PR per se) to have a `reduction` tag similar to how we have a `pointwise` tag. This was also recently a need in @Chillee's aot autograd pr.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -148,6 +149,61 @@ def getattr_recursive(obj, target):
 
+def mark_nodes_dislike_padding(g):
+    """
+    Nodes like convolution/convolution_backward want its input to be contiguous.
```
nit: s/contiguous/dense

shunting314 replied:
Hmm, what would be the difference between 'dense' and 'contiguous'? I see convolution/convolution_backward has calls like `tensor.contiguous(memory_format=the_format)`

eellison replied:
Sometimes contiguous is taken to assume `torch.contiguous(memory_format=torch.contiguous_format)`. Whereas dense is more clear it might be channels last or contiguous. It doesn't really matter, either is fine.

shunting314 replied:
Got it. Make sense to me!

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+        if len(in_strides) == 0:
+            return in_strides
+
+        # a loose check for channels last
```
Should we check whether or not `layout_opt` is True to distinguish between whether or not we are considering `size[1] == 1` as channels last or not ?

shunting314 replied:
I just refactored this code to call `Layout.is_channels_last_contiguous` as Jason suggested above. We now check for channels last more accurately.

I think this check makes sense no matter if `layout_opt` is on or off.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+
+        # get_stride_order does not work with dynamic shape. Also we can not
+        # statically decide if a padding is needed or how much padding we should
+        # do for dynamic shape.
```
Couldn't we pad to an aligned multiple even with dynamic shape ?

shunting314 replied:
We are not simply padding each stride here. We need consider stride order (check test_pad_3d_tensor for why that's the case). For dynamic shape, get_stride_order just does not work.

There may be other complexity that I have not considered yet.

Also I found we don't pad mm for dynamic shape either :)

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+        fill_order = stride_order2fill_order(stride_order)
+
+        new_strides = [0 for _ in range(len(in_strides))]
+        # since way pad when the layout is flexible, we can decide the
```
nit: /s/way/we

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+        # smallest stride to be 1.
+        new_strides[fill_order[0]] = 1
+
+        # don't align an too small stride since that cause too much memory increase.
```
nit: don't align a too small

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
+            stride = new_strides[prev_idx] * size[prev_idx]
+
+            if stride > align_stride_threshold and stride % align != 0:
+                stride = (stride + align - 1) // align * align
```
Nit: ceildiv(stride, align) * align

shunting314 committed:
Update on "[inductor] comprehensive padding"


This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] comprehensive padding"


This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] comprehensive padding"


This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] comprehensive padding"


This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


[ghstack-poisoned]

shunting314 commented:
> All perf regress has been fixed now. It turns out that the large regress from default setting from TB is due to using a new triton pin. It seems that triton regresses.. Reverting the pin change eventually fixes it. But this also means we are not using Hongtao's fix here: https://github.com/openai/triton/pull/3497 (since I need use a relatively new triton to apply the PR). I'll try to dig more.

Since Hongtao's fix may take a while to get in (still under review). And our triton pin also needs some time to be updated (even if we have all the fixes in right now, we may still need run meta internal tests. That takes time). I'll try if I can apply hongtao's fix  in the current triton pin inductor uses and measure how that works together with comprehensive padding.

eellison commented:
As a follow up would be great to see if we could use this to replace some of the current pad_mm logic.

shunting314 commented:
> As a follow up would be great to see if we could use this to replace some of the current pad_mm logic.

I think it depends on a few things
1. we can not pad weights since they have FIxedLayout. 
2. for activations, I think this PR can help pad the strides. But we'll need benchmark if padded tensor has as good matmul performance as a tensor with aligned shape.

shunting314 commented:
cc @jansel  to take another look

shunting314 committed:
Update on "[inductor] comprehensive padding"


This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang jansel Chillee eellison 


[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
